osu.Game/Screens/Select/PlaySongSelect.cs

```diff
@@ -141,17 +143,38 @@ public PlaySongSelect()
                             Width = 100,
                             Text = "Play",
                             Colour = new Color4(238, 51, 153, 255),
-                            Action = () => Push(new Player
-                            {
-                                BeatmapInfo = carousel.SelectedGroup.SelectedPanel.Beatmap,
-                                PreferredPlayMode = playMode.Value
-                            })
+                            Action = start
                         },
                     }
                 }
             };
         }
 
+        Player player;
+
+        private void start()
+        {
+            if (player != null)
+                return;
+
+            //in the future we may want to move this logic to a PlayerLoader gamemode or similar, so we can rely on the SongSelect transition
+            //and provide a better loading experience (at the moment song select is still accepting input during preload).
+            player = new Player
+            {
+                BeatmapInfo = carousel.SelectedGroup.SelectedPanel.Beatmap,
+                PreferredPlayMode = playMode.Value
+            };
+
+            player.Preload(Game, delegate
+            {
+                if (!Push(player))
+                {
+                    player = null;
+                    //error occured?

```

Comment:
Can Push fail or would Preload have failed? I feel like Push is not very error-prone, no?